portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,26 +32,43 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private ArrayList<String> messages = new ArrayList<>();
-  
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    messages.add("Hello!");
-    messages.add("Hola!");
-    messages.add("Hallo!");
+    Query query = new Query("Comments").addSort("timestamp", SortDirection.ASCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<String> commentsArrayList = new ArrayList<>();
 
-    String json = convertToJson(messages);
+    for (Entity entity : results.asIterable()) {
+      String comment = (String) entity.getProperty("comment");
+      commentsArrayList.add(comment);
+    }	
 
+    String json = convertToJsonUsingGson(commentsArrayList);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
-  private String convertToJson(ArrayList messages) {
-    String json = "[";
-    json += "{\"message\": \"" + messages.get(0) +"\"}, ";
-    json += "{\"message\": \"" + messages.get(1) +"\"}, ";
-    json += "{\"message\": \"" + messages.get(2) +"\"}";
-    json +=  "]";
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userComment = request.getParameter("comment-area");
+    long timestamp = System.currentTimeMillis();
+
+    Entity taskEntity = new Entity("Comments");
+    taskEntity.setProperty("comment", userComment);
+    taskEntity.setProperty("timestamp", timestamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(taskEntity);
+
+    response.sendRedirect("/gallery.html");
+  }
+
+  private String convertToJsonUsingGson(ArrayList arrayList) {
```

Comment:
Consider generalizing this method to take any kind of List implementation, instead of just ArrayList. It's as easy as changing the argument to (List list) and then someday if a LinkedList, for example, needed to be converted it will still work!